File: src/pretty.ml
(**プリティプリンタ：構文木を文字列に変換する*)
open Syntax

(**n回分半角4文字字下げする関数*)
let rec indent n = if n = 0 then ""
                   else indent (n - 1) ^ "    "

(**型をプリントする関数*)
let pretty_dataType = function
  | IntegerType -> "int"
  | ObjectType typeId -> typeId
  | CopyType typeId -> typeId
  | ObjectArrayType typeId -> typeId ^ "[]"
  | IntegerArrayType -> "int[]"
  | ArrayType -> ""
  | ArrayElementType -> ""
  | NilType -> ""

(**演算子をプリントする関数*)
let pretty_binOp = function
  | Add -> "+"
  | Sub -> "-"
  | Xor -> "^"
  | Mul -> "*"
  | Div -> "/"
  | Mod -> "%"
  | Band -> "&"
  | Bor -> "|"
  | And -> "&&"
  | Or -> "||"
  | Lt -> "<"
  | Gt -> ">"
  | Eq -> "="
  | Ne -> "!="
  | Le -> "<="
  | Ge -> ">="

(**代入演算子をプリントする関数*)
let pretty_modOp = function
  | ModAdd -> "+="
  | ModSub -> "-="
  | ModXor -> "^="

(**式をプリントする関数*)
let rec pretty_exp = function
  | Const n -> string_of_int n
  | Var id -> id
  | ArrayElement(id, exp) -> id ^ "[" ^ pretty_exp exp ^ "]"
  | Nil -> "nil"
  | Binary(binOp, exp1, exp2) -> pretty_exp exp1 ^ " " ^ pretty_binOp binOp ^ " " ^ pretty_exp exp2
  | Dot(exp1, exp2) -> pretty_exp exp1 ^ "." ^ pretty_exp exp2

(**変数、配列、ドット演算子をプリントする関数*)
let rec pretty_obj = function
  | VarArray(id, None) -> id
  | VarArray(id, Some exp) -> id ^ "[" ^ pretty_exp exp ^ "]"
  | InstVar(obj1, obj2) ->  pretty_obj obj1 ^ "." ^ pretty_obj obj2

(**メソッド呼び出しの引数をプリントする関数*)
let rec pretty_actArgs =
  let pretty_actArg = function
    | Id(id) -> id
    | Exp(e) -> pretty_exp e
  in
  function
  | [] -> ""
  | [arg] -> pretty_actArg arg
  | hd :: tl -> pretty_actArg hd ^ ", " ^ pretty_actArgs tl

(**文をプリントする関数：nを受け取り、n回分字下げする*)
let rec pretty_stms stms n =
  match stms with
  | [] -> ""
  | hd :: tl -> indent n ^ pretty_stm hd n ^ "\n" ^ pretty_stms tl n
  and
pretty_stm stm n =
  let s =
    match stm with
    | Assign(obj, modOp, exp) -> pretty_obj obj ^ " " ^ pretty_modOp modOp ^ " " ^ pretty_exp exp
    | Swap(obj1, obj2) -> pretty_obj obj1 ^ " <=> " ^ pretty_obj obj2
    | Conditional(exp1, stm1, stm2, exp2) -> "if " ^ pretty_exp exp1 ^ " then\n" ^ pretty_stms stm1 (n + 1) ^ indent n ^ "else\n" ^ pretty_stms stm2 (n + 1) ^ indent n ^ "fi " ^ pretty_exp exp2
    | Loop(exp1, stm1, stm2, exp2) -> "from " ^ pretty_exp exp1 ^ " do\n" ^ pretty_stms stm1 (n + 1) ^ indent n ^  "loop\n" ^ pretty_stms stm2 (n + 1) ^ indent n ^ "until " ^ pretty_exp exp2
    (* for追加部分 *)
    | For(id, e1, e2, stms) -> "for " ^ id ^ " in " ^  "(" ^ pretty_exp e1 ^ ".." ^ pretty_exp e2 ^ ")" ^ " do\n" ^ pretty_stms stms (n + 1) ^ indent n ^  "end"
      (* switch追加部分 *)
    | Switch(obj1, cases, stml, obj2) ->
       let rec pretty_cases =
         let pretty_case = function | Case -> "case" | Fcase -> "fcase" | Ecase -> "ecase" in
         let pretty_break = function | Break -> " break " | NoBreak -> "" in
         function
         | [] -> ""
         | (c, e1, s, e2, b) :: tl -> pretty_case c ^ " " ^ pretty_exp e1 ^ ":\n" ^ pretty_stms s (n + 1) ^ indent (n + 1) ^ "esac " ^ pretty_exp e2 ^ "\n" ^ indent (n + 1) ^ pretty_break b ^ "\n" ^ indent n ^ pretty_cases tl
       in
       "switch " ^ pretty_obj obj1 ^ "\n" ^ indent n ^ pretty_cases cases ^ "default:\n" ^ pretty_stms stml (n + 1) ^ indent (n + 1) ^ "break\n" ^ indent n ^ "hctiws " ^ pretty_obj obj2
    | ObjectBlock(typeId, id, stm) -> "construct " ^ typeId ^ " " ^ id ^ "\n" ^ pretty_stms stm (n + 1) ^ "\n" ^ indent n ^ "destruct " ^ id
    | LocalBlock(dataType, id, exp1, stm, exp2) -> "local " ^ pretty_dataType dataType ^ " " ^ id ^ " = " ^ pretty_exp exp1 ^ "\n" ^ pretty_stms stm n  ^ indent n ^ "delocal " ^ pretty_dataType dataType ^ " " ^ id ^ " = " ^ pretty_exp exp2
    | LocalCall(methodId, ids) -> "call " ^ methodId ^ "(" ^ pretty_actArgs ids ^ ")"
    | LocalUncall(methodId, ids) -> "uncall " ^ methodId ^ "(" ^ pretty_actArgs ids ^ ")"
    | ObjectCall(obj, methodId, ids) -> "call " ^ pretty_obj obj ^ "::" ^ methodId ^ "(" ^ pretty_actArgs ids ^ ")"
    | ObjectUncall(obj, methodId, ids) -> "uncall " ^ pretty_obj obj ^ "::" ^ methodId ^ "(" ^ pretty_actArgs ids ^ ")"
    | ObjectConstruction(typeId, obj) -> "new " ^ typeId ^ " " ^ pretty_obj obj
    | ObjectDestruction(typeId, obj) -> "delete " ^ typeId ^ " " ^ pretty_obj obj
    | CopyReference(dataType, obj1, obj2) -> "copy " ^ pretty_dataType dataType ^ " " ^ pretty_obj obj1 ^ " " ^ pretty_obj obj2
    | UncopyReference(dataType, obj1, obj2) -> "uncopy " ^ pretty_dataType dataType ^ " " ^ pretty_obj obj1 ^ " " ^ pretty_obj obj2
    | ArrayConstruction((typeId, exp), obj) -> "new " ^ typeId ^ "[" ^ pretty_exp exp ^ "] " ^ pretty_obj obj
    | ArrayDestruction((typeId, exp), obj) -> "delete " ^ typeId ^ "[" ^ pretty_exp exp ^ "] " ^ pretty_obj obj
    | Skip -> "skip"
    | Show(exp) -> "show" ^ "(" ^ pretty_exp exp ^ ")"
    | Print(str) -> "print" ^ "(\""  ^ String.escaped str ^ "\")"
    in
    s

(**フィールドまたは引数をプリントする関数*)
let pretty_decl (Decl(dataType, id)) = pretty_dataType dataType ^ " " ^ id

(**フィールドをプリントする関数(リストを受け取る)*)
let rec pretty_fields = function
  | [] -> ""
  | hd :: tl -> pretty_decl hd ^ "\n    " ^ pretty_fields tl

(**引数をプリントする関数*)
let rec pretty_args = function
  | [] -> ""
  | [arg] -> pretty_decl arg
  | hd :: tl -> pretty_decl hd ^ ", " ^ pretty_args tl

(**メソッドをプリントする関数*)
let pretty_method (MDecl(id, args, stms)) =
  "method " ^ id ^ "(" ^ pretty_args args ^ ")\n" ^ pretty_stms stms 2
(**メソッドをプリントする関数(リストを受け取る)*)
let rec pretty_methods = function
  | [m] -> pretty_method m
  | hd :: tl -> pretty_method hd ^ "\n    " ^ pretty_methods tl
  | _ -> failwith "not implemented"

(**クラスをプリントする関数*)
let pretty_c (CDecl(c, inherits, fields, methods)) =
  let inher =
  match inherits with
  | None -> ""
  | Some(id) -> " inherits " ^ id
  in
  "class " ^ c ^ inher  ^ "\n    " ^ pretty_fields fields ^ "\n    " ^ pretty_methods methods

(**クラスをプリントする関数(リストを受け取る)*)
let rec pretty_cl = function
  | [cl] -> pretty_c cl
  | hd :: tl -> pretty_c hd ^ "\n" ^ pretty_cl tl
  | _ -> failwith "not implemented"

(**プログラムをプリントする関数*)
let pretty_prog (Prog(cl)) = print_string(pretty_cl cl ^ "\n")

File: python/rooplpp/pretty.py
"""Pretty printer: AST to source code string (from pretty.ml)."""
from .syntax import *


def indent(n: int) -> str:
    return "    " * n


def pretty_datatype(dt: DataType) -> str:
    match dt:
        case IntegerType(): return "int"
        case ObjectType(tid): return tid
        case CopyType(tid): return tid
        case ObjectArrayType(tid): return tid + "[]"
        case IntegerArrayType(): return "int[]"
        case ArrayType() | ArrayElementType() | NilType(): return ""


def pretty_binop(op: BinOp) -> str:
    return {
        BinOp.Add: "+", BinOp.Sub: "-", BinOp.Xor: "^",
        BinOp.Mul: "*", BinOp.Div: "/", BinOp.Mod: "%",
        BinOp.Band: "&", BinOp.Bor: "|",
        BinOp.And: "&&", BinOp.Or: "||",
        BinOp.Lt: "<", BinOp.Gt: ">",
        BinOp.Eq: "=", BinOp.Ne: "!=",
        BinOp.Le: "<=", BinOp.Ge: ">=",
    }[op]


def pretty_modop(op: ModOp) -> str:
    return {ModOp.ModAdd: "+=", ModOp.ModSub: "-=", ModOp.ModXor: "^="}[op]


def pretty_exp(exp: Exp) -> str:
    match exp:
        case Const(n): return str(n)
        case Var(name): return name
        case ArrayElement(name, idx): return f"{name}[{pretty_exp(idx)}]"
        case Nil(): return "nil"
        case Binary(op, e1, e2):
            # 被演算子が二項演算なら括弧で囲む（pretty.ml の paren_binary と同じ）。
            # -inverse の出力を再パースできるようにするための括弧付け。
            return f"{_paren_binary(e1)} {pretty_binop(op)} {_paren_binary(e2)}"
        case Dot(e1, e2):
            return f"{pretty_exp(e1)}.{pretty_exp(e2)}"


def _paren_binary(exp: Exp) -> str:
    """二項演算の被演算子だけ括弧で囲む（添字 e[..] や Dot は不要）。"""
    if isinstance(exp, Binary):
        return "(" + pretty_exp(exp) + ")"
    return pretty_exp(exp)


def pretty_obj(obj: Obj) -> str:
    match obj:
        case VarArray(name, None): return name
        case VarArray(name, idx): return f"{name}[{pretty_exp(idx)}]"
        case InstVar(o1, o2): return f"{pretty_obj(o1)}.{pretty_obj(o2)}"


def pretty_actargs(args: list[Arg]) -> str:
    def pretty_actarg(a: Arg) -> str:
        match a:
            case IdArg(name): return name
            case ExpArg(e): return pretty_exp(e)
    return ", ".join(pretty_actarg(a) for a in args)


def pretty_stms(stms: list[Stm], n: int) -> str:
    return "".join(indent(n) + pretty_stm(s, n) + "\n" for s in stms)


def pretty_stm(stm: Stm, n: int) -> str:
    match stm:
        # 位置情報は出力しない（-inverse の出力は再パースされるため）
        case Positioned(_, s0):
            return pretty_stm(s0, n)
        case Assign(obj, op, exp):
            return f"{pretty_obj(obj)} {pretty_modop(op)} {pretty_exp(exp)}"
        case Swap(o1, o2):
            return f"{pretty_obj(o1)} <=> {pretty_obj(o2)}"
        case Conditional(e1, s1, s2, e2):
            return (f"if {pretty_exp(e1)} then\n"
                    + pretty_stms(s1, n + 1)
                    + indent(n) + "else\n"
                    + pretty_stms(s2, n + 1)
                    + indent(n) + f"fi {pretty_exp(e2)}")
        case Loop(e1, s1, s2, e2):
            return (f"from {pretty_exp(e1)} do\n"
                    + pretty_stms(s1, n + 1)
                    + indent(n) + "loop\n"
                    + pretty_stms(s2, n + 1)
                    + indent(n) + f"until {pretty_exp(e2)}")
        case For(var, e1, e2, body):
            return (f"for {var} in ({pretty_exp(e1)}..{pretty_exp(e2)}) do\n"
                    + pretty_stms(body, n + 1)
                    + indent(n) + "end")
        case Switch(obj1, cases, default, obj2):
            def pretty_case_marker(c):
                return "case" if c == Case.Case else ""
            def pretty_esac_marker(e):
                return "esac" if e == Esac.Esac else ""
            def pretty_exps(exps):
                return ":".join(pretty_exp(e) for e in exps)
            def pretty_break(b):
                return "break " if b == Break.Break else ""
            parts = ""
            for (c1, exps1), s, (c2, exps2, b) in cases:
                parts += (pretty_case_marker(c1) + " " + pretty_exps(exps1) + "\n"
                         + pretty_stms(s, n + 1)
                         + indent(n + 1) + pretty_esac_marker(c2) + " " + pretty_exps(exps2) + "\n"
                         + indent(n + 1) + pretty_break(b) + "\n"
                         + indent(n))
            return (f"switch {pretty_obj(obj1)}\n"
                    + indent(n) + parts
                    + "default\n"
                    + pretty_stms(default, n + 1)
                    + indent(n + 1) + "break\n"
                    + indent(n) + f"hctiws {pretty_obj(obj2)}")
        case ObjectBlock(tid, name, body):
            return (f"construct {tid} {name}\n"
                    + pretty_stms(body, n + 1) + "\n"
                    + indent(n) + f"destruct {name}")
        case LocalBlock(dt, name, e1, body, e2):
            return (f"local {pretty_datatype(dt)} {name} = {pretty_exp(e1)}\n"
                    + pretty_stms(body, n)
                    + indent(n) + f"delocal {pretty_datatype(dt)} {name} = {pretty_exp(e2)}")
        case LocalCall(mid, args):
            return f"call {mid}({pretty_actargs(args)})"
        case LocalUncall(mid, args):
            return f"uncall {mid}({pretty_actargs(args)})"
        case ObjectCall(obj, mid, args):
            return f"call {pretty_obj(obj)}::{mid}({pretty_actargs(args)})"
        case ObjectUncall(obj, mid, args):
            return f"uncall {pretty_obj(obj)}::{mid}({pretty_actargs(args)})"
        case ObjectConstruction(tid, obj):
            return f"new {tid} {pretty_obj(obj)}"
        case ObjectDestruction(tid, obj):
            return f"delete {tid} {pretty_obj(obj)}"
        case CopyReference(dt, o1, o2):
            return f"copy {pretty_datatype(dt)} {pretty_obj(o1)} {pretty_obj(o2)}"
        case UncopyReference(dt, o1, o2):
            return f"uncopy {pretty_datatype(dt)} {pretty_obj(o1)} {pretty_obj(o2)}"
        case ArrayConstruction(tid, exp, obj):
            return f"new {tid}[{pretty_exp(exp)}] {pretty_obj(obj)}"
        case ArrayDestruction(tid, exp, obj):
            return f"delete {tid}[{pretty_exp(exp)}] {pretty_obj(obj)}"
        case Skip():
            return "skip"
        case Show(exp):
            return f"show({pretty_exp(exp)})"
        case Print(text):
            return f'print("{_escape(text)}")'


def _escape(s: str) -> str:
    return s.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n").replace("\t", "\\t")


def pretty_decl(d: Decl) -> str:
    return f"{pretty_datatype(d.dtype)} {d.name}"


def pretty_fields(fields: list[Decl]) -> str:
    # OCaml の pretty_fields は各フィールドの後ろに "\n    " を付ける
    # （フィールドが 1 つでもあると空行が 1 行入る）。差分テストで一致させるため
    # ここも同じ形にしている。
    return "".join(pretty_decl(f) + "\n    " for f in fields)


def pretty_args(params: list[Decl]) -> str:
    return ", ".join(pretty_decl(p) for p in params)


def pretty_method(m: MDecl) -> str:
    return f"method {m.name}({pretty_args(m.params)})\n" + pretty_stms(m.body, 2)


def pretty_methods(methods: list[MDecl]) -> str:
    return "\n    ".join(pretty_method(m) for m in methods)


def pretty_class(c: CDecl) -> str:
    inher = f" inherits {c.inherits}" if c.inherits else ""
    return (f"class {c.name}{inher}\n    "
            + pretty_fields(c.fields) + "\n    "
            + pretty_methods(c.methods))


def pretty_prog(prog: Prog) -> str:
    return "\n".join(pretty_class(c) for c in prog.classes) + "\n"
